pkg/sources/huggingface/client.go

```diff
@@ -0,0 +1,214 @@
+	CreatedAt string `json:"createdAt"`
+	Data      Data   `json:"data"`
+	ID        string `json:"id"`
+}
+
+func (e Event) GetAuthor() string {
+	return e.Author.Username
+}
+
+func (e Event) GetCreatedAt() string {
+	return e.CreatedAt
+}
+
+func (e Event) GetID() string {
+	return e.ID
+}
+
+type RepoData struct {
+	FullName     string `json:"name"`
+	ResourceType string `json:"type"`
+}
+
+type Discussion struct {
+	ID        int      `json:"num"`
+	IsPR      bool     `json:"isPullRequest"`
+	CreatedAt string   `json:"createdAt"`
+	Title     string   `json:"title"`
+	Events    []Event  `json:"events"`
+	Repo      RepoData `json:"repo"`
+}
+
+func (d Discussion) GetID() string {
+	return fmt.Sprint(d.ID)
+}
+
+func (d Discussion) GetTitle() string {
+	return d.Title
+}
+
+func (d Discussion) GetCreatedAt() string {
+	return d.CreatedAt
+}
+
+func (d Discussion) GetRepo() string {
+	return d.Repo.FullName
+}
+
+// GetDiscussionPath returns the path (ex: "/models/user/repo/discussions/1") for the discussion
+func (d Discussion) GetDiscussionPath() string {
+	basePath := fmt.Sprintf("%s/%s/%s", d.GetRepo(), DiscussionsRoute, d.GetID())
+	if d.Repo.ResourceType == "model" {
+		return basePath
+	}
+	return fmt.Sprintf("%s/%s", getResourceHTMLPath(d.Repo.ResourceType), basePath)
+}
+
+// GetGitPath returns the path (ex: "/models/user/repo.git") for the repo's git directory
+func (d Discussion) GetGitPath() string {
+	basePath := fmt.Sprintf("%s.git", d.GetRepo())
+	if d.Repo.ResourceType == "model" {
+		return basePath
+	}
+	return fmt.Sprintf("%s/%s", getResourceHTMLPath(d.Repo.ResourceType), basePath)
+}
+
+type DiscussionList struct {
+	Discussions []Discussion `json:"discussions"`
+}
+
+type Repo struct {
+	IsPrivate bool   `json:"private"`
+	Owner     string `json:"author"`
+	RepoID    string `json:"id"`
+}
+
+type HFClient struct {
+	BaseURL    string
+	APIKey     string
+	HTTPClient *http.Client
+}
+
+// NewClient creates a new API client
+func NewHFClient(baseURL, apiKey string, timeout time.Duration) *HFClient {
+	return &HFClient{
+		BaseURL: baseURL,
+		APIKey:  apiKey,
+		HTTPClient: &http.Client{
+			Timeout: timeout,
+		},
+	}
+}
+
+// get makes a GET request to the Hugging Face API
+// Note: not addressing rate limit, since it seems very permissive. (ex: "If \
+// your account suddenly sends 10k requests then you’re likely to receive 503")
+func (c *HFClient) get(ctx context.Context, url string, target interface{}) error {
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create HuggingFace API request: %w", err)
+	}
+
+	req.Header.Set("Authorization", "Bearer "+c.APIKey)
+
+	resp, err := c.HTTPClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to make request to HuggingFace API: %w", err)
+	}
+	defer resp.Body.Close()
+
+	return json.NewDecoder(resp.Body).Decode(target)
```

Comment:
HuggingFace's API doesn't have the best documentation, but I believe there's a rate limit indicated by a 429 response.

https://discuss.huggingface.co/t/hugging-face-api-rate-limits/16746/6